db/migrate/20201004105659_create_doorkeeper_tables.rb

```diff
@@ -0,0 +1,53 @@
+# frozen_string_literal: true
+
+class CreateDoorkeeperTables < ActiveRecord::Migration[6.0]
+  def change
+    create_table :oauth_applications do |t|
+      t.references :owner, :null => false, :type => :bigint, :polymorphic => true
+      t.string :name, :null => false
+      t.string :uid, :null => false
+      t.string :secret, :null => false
+      t.text :redirect_uri, :null => false
+      t.string :scopes, :null => false, :default => ""
+      t.boolean :confidential, :null => false, :default => true
+      t.timestamps :null => false
+    end
+
+    add_index :oauth_applications, :uid, :unique => true
+    add_foreign_key :oauth_applications, :users, :column => :owner_id, :validate => false
+
+    create_table :oauth_access_grants do |t|
+      t.references :resource_owner, :null => false, :type => :bigint
+      t.references :application, :null => false
+      t.string :token, :null => false
+      t.integer :expires_in, :null => false
+      t.text :redirect_uri, :null => false
+      t.datetime :created_at, :null => false
+      t.datetime :revoked_at
+      t.string :scopes, :null => false, :default => ""
+      t.column :code_challenge, :string, :null => true
+      t.column :code_challenge_method, :string, :null => true
+    end
+
+    add_index :oauth_access_grants, :token, :unique => true
+    add_foreign_key :oauth_access_grants, :users, :column => :resource_owner_id, :validate => false
```

Comment:
I think we'll need a follow-on migration to validate these, or alternatively some safety_assured work given that we know the tables will be empty when the foreign key is added.

Reply:
Well it's a new table so there's nothing to validate - new records will still have to obey the constraint.

It's true that the database won't assume it's valid for all records but I'm not sure when that would matter.

I'm not really sure where that came from... It doesn't look like it came from doorkeeper's generator so I guess I added it for some reason, likely to shut safety_assured up because it didn't realise that the table was empty so complained that validation would be blocking.